File: src/common/actor/Movement.java
/**
 * @author Julian Sender
 */
package common.actor;

import java.util.HashMap;
import java.util.Set;

/**
 * Checks inputs for movements
 */
final class Movement {

    private       HashMap<String, Direction>  keymap;
    private final HashMap<Direction, Boolean> holdDown = initHoldDown();
    private       double                      velocity;

    /**
     * returns hashmap of movements
     * @return hashmap of movement-orders
     */
    private HashMap<Direction, Boolean> initHoldDown() {
        final HashMap<Direction, Boolean> hashMap = new HashMap<>();
        hashMap.put( Direction.Up, Boolean.FALSE );
        hashMap.put( Direction.Down, Boolean.FALSE );
        hashMap.put( Direction.Left, Boolean.FALSE );
        hashMap.put( Direction.Right, Boolean.FALSE );
        return hashMap;
    }

    /**
     * sets Keys
     * @param keyName UP,DOWN,LEFT,RIGHT
     * @return name of the pressed key
     */
    private boolean contains( final String keyName ) {
        return keymap.keySet().contains( keyName );
    }

    /**
     * Checks if key is held down
     * @param keyName e.g. UP,DOWN,LEFT,RIGHT
     * @return a boolean if key was hold or not
     */
    private boolean isHoldDown( final String keyName ) {
        final Direction d = this.keymap.get( keyName );
        if ( d != null ) {
            return holdDown.get( d );
        }
        return false;
    }

    /**
     *
     * @param direction instance of Direction
     * @return returns a direction
     */
    boolean isHoldDown( Direction direction ) {
        return this.holdDown.get( direction );
    }

    /**
     *
     * @param keyName Name of key pressed
     * @param b
     */
    private void setKeyHoldDownIfPresent( final String keyName, final boolean b ) {
        final Direction d = this.keymap.get( keyName );
        if ( d != null ) {
            this.holdDown.put( d, b );
        }
    }

    /**
     * Action to take when key is pressed
     * @param keyName Name of Key
     */
    void onKeyPressed( final String keyName ) {
        if ( !this.isHoldDown( keyName ) ) {
            if ( this.contains( keyName ) ) {
                this.setKeyHoldDownIfPresent( keyName, true );
            }
        }
    }

    /**
     * Action to take when key is released
     * @param keyName Keyname
     */
    void onKeyReleased( final String keyName ) {
        this.setKeyHoldDownIfPresent( keyName, false );
    }

    /**
     * GETTERS AND SETTERS
     *
     */

    Set<Direction> getDirections() {
        return this.holdDown.keySet();
    }

    double getVelocity() {
        return velocity;
    }

    void setVelocity( double i ) {
        this.velocity = i;
    }

    void setKeyMap( HashMap<String, Direction> keymap ) {
        this.keymap = keymap;
    }
}

File: Common/Actor/src/common/actor/Movement.kt
package common.actor

import javafx.scene.input.KeyCode

/**
 * Checks inputs for movements
 */
class Movement {
    lateinit var keymap: Map<KeyCode, Direction>
    val holdDown = mutableMapOf(
            Direction.Up to false,
            Direction.Down to false,
            Direction.Left to false,
            Direction.Right to false)
    var velocity = 0.0

    /**
     * Checks if key is held down
     * @param keyName e.g. UP,DOWN,LEFT,RIGHT
     * @return a boolean if key was hold or not
     */
    private fun isHoldDown(keyName: KeyCode): Boolean {
        val d = keymap[keyName]
        return if (d != null) {
            holdDown[d]!!
        } else false
    }

    /**
     *
     * @param keyName Name of key pressed
     * @param b
     */
    private fun setKeyHoldDownIfPresent(keyName: KeyCode, b: Boolean) {
        val d = keymap[keyName]
        if (d != null) {
            holdDown[d] = b
        }
    }

    /**
     * Action to take when key is pressed
     * @param keyName Name of Key
     */
    fun onKeyPressed(keyName: KeyCode) {
        if (!this.isHoldDown(keyName) && keymap.containsKey(keyName)) {
            setKeyHoldDownIfPresent(keyName, true)
        }
    }

    /**
     * Action to take when key is released
     * @param keyName Keyname
     */
    fun onKeyReleased(keyName: KeyCode) = setKeyHoldDownIfPresent(keyName, false)


}
